src/core/NgenSimulation.cpp

```diff
@@ -0,0 +1,253 @@
+        catchment_outflows_.resize(catchment_outflows_.size() + catchment_indexes_.size());
+        nexus_downstream_flows_.resize(nexus_downstream_flows_.size() + nexus_indexes_.size());
+
+        // this is the time that layers are trying to reach (or get as close as possible)
+        auto next_time = catchment_formulation_manager_->Simulation_Time_Object->next_timestep_epoch_time();
+        auto next_time_private = sim_time_->next_timestep_epoch_time();
+
+        // this is the time that the layer above the current layer is at
+        auto prev_layer_time = next_time;
+
+        // this is the time that the least advanced layer is at
+        auto layer_min_next_time = next_time;
+        do {
+            for (auto& layer : layers_) {
+                auto layer_next_time = layer->next_timestep_epoch_time();
+
+                // only advance if you would not pass the master next time and the previous layer
+                // next time
+                if (layer_next_time <= next_time && layer_next_time <= prev_layer_time) {
+                    if (simulation_step_ % 100 == 0) {
+                        ss << "Updating layer: '" << layer->get_name() << "' at output step " << simulation_step_;
+                        LOG(ss.str(), LogLevel::DEBUG);
+                        ss.str("");
+                    }
+#if NGEN_WITH_ROUTING
+                    boost::span<double> catchment_span(catchment_outflows_.data() + (simulation_step_ * catchment_indexes_.size()),
+                                                       catchment_indexes_.size());
+                    boost::span<double> nexus_span(nexus_downstream_flows.data() + (simulation_step_ * nexus_indexes_.size()),
+                                                   nexus_indexes_.size());
+#else
+                    boost::span<double> catchment_span;
+                    boost::span<double> nexus_span;
+#endif
+                    layer->update_models(
+                        catchment_span,
+                        catchment_indexes_,
+                        nexus_span,
+                        nexus_indexes_,
+                        simulation_step_
+                    ); // assume update_models() calls time->advance_timestep()
+                    prev_layer_time = layer_next_time;
+                } else {
+                    layer_min_next_time = prev_layer_time = layer->current_timestep_epoch_time();
+                }
+
+                if (layer_min_next_time > layer_next_time) {
+                    layer_min_next_time = layer_next_time;
+                }
+            } // done layers
+        } while (layer_min_next_time < next_time); // rerun the loop until the last layer would pass the master next time
+
+        if (simulation_step_ + 1 < num_times) {
+            sim_time_->advance_timestep();
+            catchment_formulation_manager_->Simulation_Time_Object->advance_timestep();
+        }
+    }
+}
+
+
+void NgenSimulation::advance_one_output_step()
```

Comment:
My intention is to move the body of the above loop down here, so that we can separate out any control logic for output and state saving.